### ISSUE TITLE
False positive type inferencing with "and" operator

### ISSUE THREAD

**Issue by award999** [microsoft/pyright#3024]
**Describe the bug**
Pyright seems to be making a wrong type inference when assigning a value using the "and" operator

**To Reproduce**
1. If you run the code below you should see "Processed 5." is printed, so I believe the type of "data" should her inferred to be "str | None"
2. Running pyright against this file throws an error since it has inferred the the type is "int | str | None"

**Expected behavior**
Infer "str | None"

**Screenshots or Code**
```python
`from typing import NamedTuple, Optional

def load_data() -> Optional[int]:
    # In a real examples this function would return `None` if the data does not exist.
    return 5

def process(data: int) -> str:
    return f"Processed {data}."

maybe_data = load_data()
# If the data exists we want to process it:
data = maybe_data and process(maybe_data)

class Foo(NamedTuple):
    processed_data: Optional[str]

foo = Foo(processed_data=data)  # < pyright reports an error here
print(foo.processed_data)
# Argument of type "int | str | None" cannot be assigned to parameter "processed_data" of type "str | None" in function "__init__"
#   Type "int | str | None" cannot be assigned to type "str | None"
#     Type "int" cannot be assigned to type "str | None"
#       "int" is incompatible with "str"
#       Type cannot be assigned to type "None"PyrightreportGeneralTypeIssues`
```

**VS Code extension or command-line**
Running from command line, version 1.1.220

**Comment by erictraut** [comment]
The `and` operator returns the left-hand operand if it evaluates to falsy. An `int` value can evaluate to falsy if it is zero. If `maybe_data` is zero, then `maybe_data and process(maybe_data)` will evaluate to zero, so it's not safe for the type narrowing logic to remove `int` from the union in this case.

**Comment by award999** [comment]
that's makes sense, thanks
